File: bin/ruby-cli/bin/investigate-naval-js.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
# @category: dev
# @description: Investigate nav.al JavaScript and Read More functionality
# @tags: testing, web-scraping, debugging

require 'ferrum'

puts "🔍 Investigating nav.al JavaScript and Read More functionality..."

browser = Ferrum::Browser.new(headless: false, timeout: 60)

begin
  puts "📄 Loading https://nav.al..."
  browser.goto('https://nav.al')
  sleep(5) # Wait for all JavaScript to load

  # Check if jQuery or other libraries are loaded
  puts "\n📚 Checking JavaScript libraries..."
  jquery_version = browser.evaluate("typeof jQuery !== 'undefined' ? jQuery.fn.jquery : 'not loaded'") rescue 'error'
  puts "  jQuery: #{jquery_version}"

  dollar_available = browser.evaluate("typeof $ !== 'undefined'") rescue false
  puts "  $ available: #{dollar_available}"

  # Check what JavaScript events might be attached
  puts "\n🔧 Investigating Read More button..."

  # Find the specific button
  trigger_element = browser.css('.trigger-load-more').first

  if trigger_element
    puts "  ✅ Found trigger-load-more element"

    # Check its properties
    href = trigger_element.attribute('href') rescue ""
    classes = trigger_element.attribute('class') rescue ""
    onclick = trigger_element.attribute('onclick') rescue ""

    puts "    Href: #{href}"
    puts "    Classes: #{classes}"
    puts "    OnClick: #{onclick.empty? ? 'none' : onclick}"

    # Check if it has jQuery event handlers
    has_click_handlers = browser.evaluate("
      var element = arguments[0];
      var events = $._data ? $._data(element, 'events') : null;
      return events && events.click ? events.click.length : 0;
    ", trigger_element) rescue 0

    puts "    jQuery click handlers: #{has_click_handlers}"

    # Check for data attributes that might indicate AJAX behavior
    data_attrs = browser.evaluate("
      var element = arguments[0];
      var data = {};
      for (var i = 0; i < element.attributes.length; i++) {
        var attr = element.attributes[i];
        if (attr.name.startsWith('data-')) {
          data[attr.name] = attr.value;
        }
      }
      return data;
    ", trigger_element) rescue {}

    puts "    Data attributes: #{data_attrs}"

    # Check what happens when we scroll to the element
    puts "\n🔄 Scrolling to Read More button..."
    trigger_element.evaluate("this.scrollIntoView({behavior: 'smooth', block: 'center'})")
    sleep(2)

    # Check if the button becomes different after scrolling
    visible_after_scroll = trigger_element.displayed? rescue false
    puts "    Visible after scroll: #{visible_after_scroll}"

    # Try to trigger any scroll-based loading
    puts "\n📜 Testing scroll-based loading..."
    initial_height = browser.evaluate("document.body.scrollHeight")
    browser.evaluate("window.scrollTo(0, document.body.scrollHeight)")
    sleep(3)

    new_height = browser.evaluate("document.body.scrollHeight")
    puts "    Scroll height: #{initial_height} → #{new_height}"

    if new_height > initial_height
      puts "    🎉 Infinite scroll detected!"

      # Count links again
      new_links = browser.css('a[href]').size
      puts "    New link count: #{new_links}"
    else
      puts "    No infinite scroll detected"
    end

    # Try clicking with more aggressive event simulation
    puts "\n🎯 Testing aggressive click simulation..."

    # First try: Multiple event types
    result1 = browser.evaluate("
      var element = arguments[0];
      var events = ['mousedown', 'mouseup', 'click'];
      var results = [];

      events.forEach(function(eventType) {
        var event = new MouseEvent(eventType, {
          view: window,
          bubbles: true,
          cancelable: true,
          button: 0
        });
        var dispatched = element.dispatchEvent(event);
        results.push(eventType + ':' + dispatched);
      });

      return results.join(', ');
    ", trigger_element) rescue 'error'

    puts "    Mouse events result: #{result1}"
    sleep(3)

    # Check for any network activity or DOM changes
    link_count_after = browser.css('a[href]').size
    puts "    Links after mouse events: #{link_count_after}"

    # Second try: Focus and keyboard events
    result2 = browser.evaluate("
      var element = arguments[0];
      element.focus();

      var enterEvent = new KeyboardEvent('keydown', {
        key: 'Enter',
        code: 'Enter',
        keyCode: 13,
        bubbles: true,
        cancelable: true
      });

      return element.dispatchEvent(enterEvent);
    ", trigger_element) rescue 'error'

    puts "    Keyboard event result: #{result2}"
    sleep(3)

    link_count_after2 = browser.css('a[href]').size
    puts "    Links after keyboard events: #{link_count_after2}"

    # Third try: Check if there are any AJAX requests we can trigger manually
    puts "\n🌐 Checking for AJAX endpoints..."

    # Look for any fetch or XMLHttpRequest patterns in the page source
    page_source = browser.evaluate("document.documentElement.outerHTML")

    ajax_patterns = [
      /fetch\s*\(\s*['"](.*?)['"]/,
      /XMLHttpRequest.*open\s*\(\s*['"]GET['"],\s*['"](.*?)['"]/,
      /\$\.get\s*\(\s*['"](.*?)['"]/,
      /\$\.ajax.*url\s*:\s*['"](.*?)['"]/
    ]

    ajax_urls = []
    ajax_patterns.each do |pattern|
      matches = page_source.scan(pattern)
      ajax_urls.concat(matches.flatten) if matches.any?
    end

    if ajax_urls.any?
      puts "    Found potential AJAX URLs:"
      ajax_urls.uniq.each { |url| puts "      #{url}" }
    else
      puts "    No obvious AJAX patterns found"
    end

    # Fourth try: Check for any load-more specific JavaScript functions
    load_more_functions = browser.evaluate("
      var functions = [];
      for (var prop in window) {
        if (typeof window[prop] === 'function' &&
            (prop.toLowerCase().includes('load') ||
             prop.toLowerCase().includes('more') ||
             prop.toLowerCase().includes('next'))) {
          functions.push(prop);
        }
      }
      return functions;
    ") rescue []

    if load_more_functions.any?
      puts "    Load-more related functions: #{load_more_functions.join(', ')}"

      # Try calling them
      load_more_functions.each do |func|
        begin
          puts "    Trying to call #{func}()..."
          browser.evaluate("if (typeof #{func} === 'function') #{func}();")
          sleep(2)

          link_count_after_func = browser.css('a[href]').size
          puts "      Links after #{func}: #{link_count_after_func}"
        rescue => e
          puts "      #{func} failed: #{e.message}"
        end
      end
    else
      puts "    No load-more functions found in global scope"
    end

  else
    puts "  ❌ trigger-load-more element not found"
  end
ensure
  puts "\n🔒 Closing browser..."
  browser.quit rescue nil
end

puts "✅ Investigation complete!"

File: bin/bun-cli/src/scripts/browser/investigate-naval-js.ts
import { Script } from "../../core/decorators/Script";
import { Script as ScriptBase } from "../../core/base/Script";
import type { Context } from "../../core/types";
import { BrowserService } from "../../core/services/BrowserService";

@Script({
  emoji: "🔍",
  tags: ["browser", "debug", "naval"],
  args: {
    url: {
      type: "string",
      position: 0,
      required: false,
      default: "https://nav.al",
      description: "URL to inspect",
    },
  },
})
export class InvestigateNavalJs extends ScriptBase {
  async run(ctx: Context): Promise<void> {
    const url = ctx.args.url as string;
    const browser = new BrowserService({ logger: this.logger });

    this.logger.banner("Investigate nav.al JavaScript");
    const initialized = await browser.initialize();
    if (!initialized) {
      throw new Error("Browser service unavailable. Install puppeteer to enable this command.");
    }

    try {
      const report = await browser.withPage(async (page) => {
        await page.goto(url, { waitUntil: "networkidle0", timeout: 60000 });
        await page.waitForTimeout(5000);

        const beforeHeight = await page.evaluate(() => document.body.scrollHeight);
        const initialLinks = await page.$$eval("a[href]", (links: Element[]) => links.length);
        const triggerPresent = await page.$(".trigger-load-more");

        let triggerDetails = null;
        if (triggerPresent) {
          triggerDetails = await page.$eval(".trigger-load-more", (element: Element) => ({
            text: element.textContent?.trim() || "",
            className: (element as HTMLElement).className || "",
            href: (element as HTMLAnchorElement).getAttribute("href") || "",
            onclick: (element as HTMLAnchorElement).getAttribute("onclick") || "",
            dataAttributes: Array.from(element.attributes)
              .filter((attribute) => attribute.name.startsWith("data-"))
              .reduce<Record<string, string>>((acc, attribute) => {
                acc[attribute.name] = attribute.value;
                return acc;
              }, {}),
          }));

          await page.$eval(".trigger-load-more", (element: Element) => {
            (element as HTMLElement).scrollIntoView({ behavior: "auto", block: "center" });
          });
          await page.waitForTimeout(1500);
        }

        await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
        await page.waitForTimeout(3000);

        const afterHeight = await page.evaluate(() => document.body.scrollHeight);
        const afterScrollLinks = await page.$$eval("a[href]", (links: Element[]) => links.length);

        let clickEffects = {
          mouseEvents: "",
          keyboardEvent: false,
          linkCountAfterClick: afterScrollLinks,
          potentialFunctions: [] as string[],
        };

        if (triggerPresent) {
          clickEffects = await page.$eval(".trigger-load-more", (element: Element) => {
            const target = element as HTMLElement;
            const mouseEvents = ["mousedown", "mouseup", "click"]
              .map((type) => {
                const event = new MouseEvent(type, {
                  bubbles: true,
                  cancelable: true,
                  view: window,
                  button: 0,
                });
                return `${type}:${target.dispatchEvent(event)}`;
              })
              .join(", ");

            target.focus();
            const keyboardEvent = target.dispatchEvent(
              new KeyboardEvent("keydown", {
                key: "Enter",
                code: "Enter",
                keyCode: 13,
                bubbles: true,
                cancelable: true,
              })
            );

            const potentialFunctions = Object.keys(window).filter((key) => {
              const value = (window as Record<string, unknown>)[key];
              return (
                typeof value === "function" &&
                ["load", "more", "next"].some((token) => key.toLowerCase().includes(token))
              );
            });

            return {
              mouseEvents,
              keyboardEvent,
              linkCountAfterClick: document.querySelectorAll("a[href]").length,
              potentialFunctions,
            };
          });
        }

        const libraries = await page.evaluate(() => ({
          jqueryVersion:
            typeof (window as any).jQuery !== "undefined" ? (window as any).jQuery.fn.jquery : "not loaded",
          dollarAvailable: typeof (window as any).$ !== "undefined",
        }));

        return {
          url,
          initialLinks,
          beforeHeight,
          afterHeight,
          afterScrollLinks,
          triggerDetails,
          libraries,
          clickEffects,
        };
      });

      console.log(`URL: ${report.url}`);
      console.log(`jQuery: ${report.libraries.jqueryVersion}`);
      console.log(`$ available: ${report.libraries.dollarAvailable}`);
      console.log(`Initial links: ${report.initialLinks}`);
      console.log(`Scroll height: ${report.beforeHeight} -> ${report.afterHeight}`);
      console.log(`Links after scroll: ${report.afterScrollLinks}`);
      if (report.triggerDetails) {
        console.log("Trigger details:");
        console.log(`  Text: ${report.triggerDetails.text}`);
        console.log(`  Class: ${report.triggerDetails.className}`);
        console.log(`  Href: ${report.triggerDetails.href || "none"}`);
        console.log(`  OnClick: ${report.triggerDetails.onclick || "none"}`);
        console.log(`  Data: ${JSON.stringify(report.triggerDetails.dataAttributes)}`);
        console.log(`Mouse events: ${report.clickEffects.mouseEvents}`);
        console.log(`Keyboard event: ${report.clickEffects.keyboardEvent}`);
        console.log(`Links after click simulation: ${report.clickEffects.linkCountAfterClick}`);
        console.log(`Potential global functions: ${report.clickEffects.potentialFunctions.join(", ") || "none"}`);
      } else {
        console.log("No .trigger-load-more element found.");
      }
    } finally {
      await browser.close();
    }
  }
}
